src/detect-quic-version.c

```diff
@@ -0,0 +1,264 @@
+        DetectEngineThreadCtx *det_ctx, const Signature *s, const SigMatchData *smd, Flow *f,
+        uint8_t flags, void *alstate, void *txv, uint64_t tx_id);
+
+/**
+ * \brief Registration function for quic.version: keyword
+ */
+void DetectQuicVersionRegister(void)
+{
+    sigmatch_table[DETECT_AL_QUIC_VERSION].name = "quic.version";
+    sigmatch_table[DETECT_AL_QUIC_VERSION].desc = "match Quic version";
+    sigmatch_table[DETECT_AL_QUIC_VERSION].url = "/rules/quic-keywords.html#quic-version";
+    sigmatch_table[DETECT_AL_QUIC_VERSION].AppLayerTxMatch = DetectQuicVersionMatch;
+    sigmatch_table[DETECT_AL_QUIC_VERSION].Setup = DetectQuicVersionSetup;
+    sigmatch_table[DETECT_AL_QUIC_VERSION].Free = DetectQuicVersionFree;
+#ifdef UNITTESTS
+    sigmatch_table[DETECT_AL_QUIC_VERSION].RegisterTests = QuicVersionRegisterTests;
+#endif
+
+    DetectAppLayerInspectEngineRegister("quic.version", ALPROTO_QUIC, SIG_FLAG_TOSERVER, 1,
+            DetectEngineInspectQuicVersionGeneric);
+
+    quic_version_id = DetectBufferTypeGetByName("quic.version");
+}
+
+static int DetectEngineInspectQuicVersionGeneric(ThreadVars *tv, DetectEngineCtx *de_ctx,
+        DetectEngineThreadCtx *det_ctx, const Signature *s, const SigMatchData *smd, Flow *f,
+        uint8_t flags, void *alstate, void *txv, uint64_t tx_id)
+{
+    return DetectEngineInspectGenericList(
+            tv, de_ctx, det_ctx, s, smd, f, flags, alstate, txv, tx_id);
+}
+
+/**
+ * \internal
+ * \brief Function to match protocol version of an Quic Tx
+ *
+ * \param det_ctx Pointer to the pattern matcher thread.
+ * \param f       Pointer to the current flow.
+ * \param flags   Flags.
+ * \param state   App layer state.
+ * \param txv     Pointer to the transaction.
+ * \param s       Pointer to the Signature.
+ * \param ctx     Pointer to the sigmatch that we will cast into DetectQuicVersionData.
+ *
+ * \retval 0 no match.
+ * \retval 1 match.
+ */
+static int DetectQuicVersionMatch(DetectEngineThreadCtx *det_ctx, Flow *f, uint8_t flags,
+        void *state, void *txv, const Signature *s, const SigMatchCtx *ctx)
+{
+    const DetectU32Data *de = (const DetectU32Data *)ctx;
+    uint32_t version;
+
+    version = rs_quic_tx_get_version(state);
+    SCLogNotice("QUIC: %d", version);
+
+    return DetectU32Match(version, de);
+}
+
+/**
+ * \internal
+ * \brief this function is used to add the parsed sigmatch  into the current signature
+ *
+ * \param de_ctx pointer to the Detection Engine Context
+ * \param s pointer to the Current Signature
+ * \param rawstr pointer to the user provided options
+ *
+ * \retval 0 on Success
+ * \retval -1 on Failure
+ */
+static int DetectQuicVersionSetup(DetectEngineCtx *de_ctx, Signature *s, const char *rawstr)
+{
+    SigMatch *sm = NULL;
+    DetectU32Data *de = NULL;
+
+    if (DetectSignatureSetAppProto(s, ALPROTO_QUIC) < 0)
+        return -1;
+
+    de = DetectU32Parse(rawstr);
```

Comment:
Should there be some more magic for special constants such as `QuicVersion::Q043` ?
Do they have an official name ?

Reply:
In the protocol versions are specified as a u32, they can have an ascii representation such as Q043

Reply:
> they can have an ascii representation such as Q043

Do we want to be able to use this ascii representation in signatures ?